### ISSUE TITLE
Incremental provision real time trace problems

### ISSUE THREAD

**Issue by sergeisnitko** [SubPointSolutions/spmeta2#978]
Hello, I have some problems with real time trace in incremental provision

Version: SPMeta2.CSOM.Standard 1.2.120-beta2

On _OnModelNodeProcessed_ event I try to show the information about state of incremental provision (should it be deployed or not) and every time I get that the node don't need to be deployed. But after full deployment I can get the state of incremental provision of each node and it is correct. So the incremental deploy works correct, it works not so fine on _OnModelNodeProcessed_ event

Some screenshots for more information:
![screenshot_8](https://cloud.githubusercontent.com/assets/7835251/23950232/529b0188-099b-11e7-8c45-b9cee21fb8cf.png)

![img_15032017_154856_0](https://cloud.githubusercontent.com/assets/7835251/23950177/22245cac-099b-11e7-968c-3dbe5123172a.png)

![img_15032017_155011_0](https://cloud.githubusercontent.com/assets/7835251/23950191/2af59a6c-099b-11e7-94ca-a1b276b472c2.png)

The sample model:
```c#
var SiteAssetsModel = SPMeta2Model.NewSiteModel(site =>
{
	site
		.AddUserCustomAction(Core.SpfActions.SpfSearchHackAction())
		.AddUserCustomAction(Core.SpfActions.SpfShortLinksAction())
		.AddUserCustomAction(Core.SpfActions.SpfMPArtefactsAction())

		.AddSiteFeature(Core.SpfFeatures.SharePointServerPublishingInfrastructure)

		.AddTaxonomyTermStore(SpfTaxonomy.DefaultSiteTermStore(), termStore =>
		{
			termStore
				.AddTaxonomyTermGroup(SpfTaxonomy.MediaBankTermGroup(), group =>
				{
					group
						.AddTaxonomyTermSet(SpfTaxonomy.MediaBankCategoryTermSet())
						.AddTaxonomyTermSet(SpfTaxonomy.MediaBankKeywordsTermSet())
						.AddTaxonomyTermSet(SpfNavigation.GlobalNavigation(), navigation=>
						{
							navigation
								.AddNavigationTerms()
								;
						})
						;
				})
				;
		})
		.AddRootWeb(new RootWebDefinition(), RootWeb =>
		{
			RootWeb
				.AddWebNavigationSettings(SpfNavigation.WebGlobalNavigation())
				.AddHostList(BuiltInListDefinitions.Catalogs.MasterPage, list =>
				{
					var FolderPath = Path.Combine(SpfExtentions.SystemPath, SpfExtentions.DinamicAssets);
					if (Directory.Exists(FolderPath))
					{
						SpfExtentions.FillApplytoFile();
						ModuleFileUtils.LoadModuleFilesFromLocalFolder(list, FolderPath);
					}

				})
				.AddWeb(SpfWebs.ShortLinksWeb(), InWeb =>
				{
					InWeb
						.AddWelcomePage(new WelcomePageDefinition
						{
							Url = UrlUtility.CombineUrl(BuiltInListDefinitions.Pages.CustomUrl, SpfPublishPages.ShortLinkPage().FileName)
						})

						.AddMasterPageSettings(SpfPages.MasterPage())

						.AddWebFeature(Core.SpfFeatures.DisableMinimalDownloadStrategy)
						.AddWebFeature(Core.SpfFeatures.SharePointServerPublishing)

						.AddList(SpfPublishPages.Pages(), SitePages =>
						 {
							 SitePages
								.AddPublishingPage(SpfPublishPages.ShortLinkPage(), SPage =>
								{
									SPage
										.AddWebPart(Core.SpfPages.HideLeftNavWP())
										.AddWebPart(Core.SpfPages.ShortLinkCEWP());
								});

						});
				})
				;

		});
});
SiteAssetsModel.SetIncrementalProvisionModelId("Assets");
```

The deploy code is here:
```c#
public static void DeployModel(this ClientContext Ctx, WebModelNode model)
{
	BeforeDeployModel(x =>
	{
		x.DeployModel(SPMeta2.CSOM.ModelHosts.WebModelHost.FromClientContext(Ctx), model);
	});
}
public static void DeployModel(this ClientContext Ctx, SiteModelNode model)
{
	BeforeDeployModel(x =>
	{
		x.DeployModel(SPMeta2.CSOM.ModelHosts.SiteModelHost.FromClientContext(Ctx), model);
	});

}
public static void BeforeDeployModel(Action<StandardCSOMProvisionService> Deploy)
{
	var provisionService = new StandardCSOMProvisionService();

	var IncProvisionConfig = new IncrementalProvisionConfig();
	IncProvisionConfig.AutoDetectSharePointPersistenceStorage = true;
	provisionService.SetIncrementalProvisionMode(IncProvisionConfig);
	provisionService.OnModelNodeProcessed += ModelNodeProcessed;
	Deploy(provisionService);
	provisionService.SetDefaultProvisionMode();
	Console.WriteLine();
	Console.WriteLine();
}

public static void ModelNodeProcessed(object sender, ModelProcessingEventArgs args)
{
	var ModelId = args.Model.GetPropertyBagValue(DefaultModelNodePropertyBagValue.Sys.IncrementalProvision.PersistenceStorageModelId);

	bool shouldDeploy = args.CurrentNode.GetIncrementalRequireSelfProcessingValue();

	var NodeName = args.CurrentNode.Value.ToString();
	if (NodeName.Length > 20)
	{
		NodeName = NodeName.Substring(0, 20) + "...";
	}

	Console.WriteLine(
		string.Format("{5}[{6}] [{0}/{1}] - [{2}%] - [{3}] [{4}]",
		new object[] {
			args.ProcessedModelNodeCount.AddZeros(3),
			args.TotalModelNodeCount.AddZeros(3),
			Math.Round(100d * (double)args.ProcessedModelNodeCount / (double)args.TotalModelNodeCount).AddSpacesBefore(3),
			args.CurrentNode.Value.GetType().Name,
			NodeName,
			(shouldDeploy == true) ? "[+]" : "[-]",
			ModelId
	   }));
}
```

**Comment by SubPointSupport** [comment]
Hey @sergeisnitko, thanks for reporting this.

Happy to see incremental provision working for you. Seems to be a minor issue, we'll have a deeper look.

**Comment by SubPointSupport** [comment]
Okay confirming the issue. We need to call OnAfterDeployModelNode before OnModelProcessed. OnAfterDeployModelNode methos actually gets overwritten for incremental provision, so that it updates the model node state in non-persistent property bag. Will fix that before releaing major 1.2.120 version.
```cs
if (OnModelProcessed != null)
    OnModelProcessed(modelNode);

OnAfterDeployModelNode(modelHost, modelNode);
```
https://github.com/SubPointSolutions/spmeta2/blob/dev/SPMeta2/SPMeta2/Services/Impl/DefaultModelTreeTraverseService.cs#L129

**Comment by SubPointSupport** [comment]
And really thank you for trying out incremental provision and providing feedback really appreciate that!
